### ISSUE TITLE
Make abstract classes "Layouts" and rename "LuckyWeb::Page" to "LuckyWeb::Html"

### ISSUE THREAD

**Issue by paulcsmith** [luckyframework/web#185]
Instead of having separate page and layout classes, just combine them.

```crystal
# Define a page to inherit from
abstract class PageWithSidebar < LuckyWeb::Page
  def render
    html do
      head do
        title page_title
      end

      body do
        # Call the inner method to render the page contents
        inner
        div class: "sidebar" do
          sidebar
        end
      end
    end
  end

  abstract def inner
  abstract def sidebar

  def page_title
    "Default title"
  end
end

class Users::IndexPage < PageWithSidebar
  def inner
    h1 "This renders inside of the main page"
  end

  def sidebar
    ul do
      li "In the sidebar"
    end
  end

  def page_title
    "List of users"
  end
end
```

**Comment by paulcsmith** [comment]
@edwardloveall I was talking with @SViccari about the layouts in the guides and after some thought realized that an abstract page is always tied to a specific layout. So rather than splitting layouts/pages, I could change it to just combine the concepts.

What do you think of this approach? Does this make more sense than what was in the guides?

**Comment by edwardloveall** [comment]
I'll have to think about this for a bit. My grasp on pages is pretty weak still, but I'll re-read the guide and let you know!

**Comment by SViccari** [comment]
TL;DR I like this idea.

Oooh, very interesting idea. I find it intriguing that Phoenix employs this strategy (rendering HTML from the view module via `render/2`) and yet, decided to keep templates and views separate. Perhaps it's because they wanted to use HTML with EEx, rather than build something akin to `BaseTags`, encouraging devs to define all HTML in the view/page.

When it comes to rendering HTML within HTML, Phoenix and Rails achieves this by defining a base layout, which can be overridden from the controller. Which, as you pointed out, is a strange responsibility for the controller. With this approach, the responsibility resides with the Page object and rendering HTML within HTML is achieved through inheritance.

I agree that combining `layout` and `page` into one concept feels more accurate and it seems like a nice simplification 👍.

**Comment by edwardloveall** [comment]
TL;DR I also like this, but I have a bunch of other thoughts and they get complicated. 😛 

I think this is much better that the current implementation. It's cleaner and easier to learn/remember. You both make a great point about how it's better for the Page to be responsible for the layout instead of the controller.

---

I'm trying to take a step back and figure how Lucky's take on views, called Pages. A Page's job is to make sure:

* Elements required for any given page are included
* Page generation stays DRY/doesn't get tedious

Is that right? The more I think about these goals the trickier it gets.

In general, you have three types of content:

1. Content directly relevant to the page (list of users, sign up form, etc)
2. Auxiliary content (sidebar, header, footer, etc)
3. Non-visible meta content (`<head>`, `<DOCTYPE>`, stylesheet link tags, etc)

Seems like you've covered 1 well so far in Lucky, but maybe this approach is trying to combine 2 and 3 when it's not appropriate. They can both be considered "shared code", but _usually_ 2 needs some support from an action, and 3 doesn't. I think that's what's making it tricky for me: shared code, but not shared support. I thought about for a while and came up with couple of hastily-typed ideas below.

---

I'm almost thinking that including modules would be better than inheriting from a base class. Consider that this page now needs some kind of global ticker tape. Should it inherit from a `TickerTapeWithSidebarPage`? Should there just be a ticker tape partial? Or should `TickerTapePage` become the base class and sidebar become the partial? I'm not sure where to draw the line. Maybe I'm getting hung up on the sidebar example though. 🙂 

Here's what a composition (including modules) based approach could look like:

```crystal
module PageWithSidebar
  abstract def sidebar
end

class Users::IndexPage
  include PageWithSidebar

  def sidebar
    # ...
  end
end
```

Downside with this is doesn't put sidebar anywhere on the page in particular, it just ensures that you have one.

---

Totally spitballing here: what if there were actions that were not connected to a resource, like a Sidebar::Show action that you could pass things to? You could guarantee type safety and include it anywhere in the page you wanted:

```crystal
render do
  render_announcements
  render_dashboard_stats
  present Sidebar::Show.new(current_page: @page)
end
```

Then the `Sidebar::Show` action is completely responsible for fetching things from a database and displaying its own page, like a partial. I'm sure this has its own set of problems though. For example, you still may need to ensure that the base action has whatever context for this sub-action.

---

Anyway, I think I've gone in too deep on this and sufficiently talked myself in circles. If any of them grab you, let me know. I won't be offended _at all_ if you just go with your original suggestion for now and ignore these. This helped me think it through though. 🙂

**Comment by paulcsmith** [comment]
@edwardloveall Those are some awesome thoughts. I think your example with the TickerTape definitely brings up some interesting ideas. 

To clarify things I think I'll extract a new guide that talks about layouts and components in order to help people not just use them, but figure out *when* modules/classes should be used.

I also think I didn't do a good job explaining when I'd extract a `PageWithSidebar`. I would only do that if a whole bunch of pages all have a sidebar and the layout changes significantly (maybe a different header and footer along with it). In some cases (like the ticker tap example) you wouldn't want to extract a whole new base class or you might slowly end up with `PageWithSidebarAndTickerTapeAndFooterAndStuff`. Which would be AWFUL.

So I think the module approach and the class inheritance approach are both helpful, but it needs to be explained well. 

As an example, if I had a bunch of pages with a sidebar and other stuff I'd have a `PageWithSidebar`, and if I wanted it to also be *capable* of having ticker tape every now and then, I'd do this:

```
class PageWithSidebar
  render do
    # render the main contents
    inner
    # render the sidebar
    sidebar
    # render ticker_tape if the class implements it
    if responds_to?(:ticker_tape)
      ticker_tape
    end
  end

  abstract def sidebar
  abstract def inner
end

class Users::Index < PageWithSidebar
  # Optionally create a PageWithTickerTape module
  # if you want to have an `abstract def` and maybe some other methods used for 
  # working with ticker tape HTML like a `render_ticker_item` method

  def inner
    h1 "List of users"
  end

  def sidebar
    h2 "Sidebar stuff"
  end

  # if you want ticker tape to show up
  def ticker_tape
    h3 "Ticker tape"
  end
end
```

Does that make sense? If so, I think the guides need to take about when it's best to extract a new class (when lots of pages use the layout and they are significantly different than other pages) and when you should use `responds_to?` to allow pages to optionally show things in the layout.

I also love the idea of extracting classes, and that can be done right now. I just haven't documented it. 

Here's an example:

```
render do
  Users::ListComponent.new(self, users: @users).render
end

class Users::ListComponent
  forward_missing_to @page

  def initialize(@page : LuckyWeb::Page, @users : UserQuery)
  end

  def render
    @users.each do |user|
      ul do
        li user.name
      end
    end
  end
end
```

In practice I would use modules to share HTML that doesn't have a lot of conditionals or state. It's easy to just have a single method that renders some HTML. If the HTML is more complex, then I'd extract a component class that handle the conditionals in private methods.

What do you think? BTW, I really appreciate your thoughts. It's very helpful!

**Comment by paulcsmith** [comment]
And yes the goals of the page are to represent the full HTML in a way that is DRY, easy to share, and type safe. I think the approach of combining classes for significantly different pages, and `responds_to?` for optional pieces within a page makes for a pretty good combo. But that could be my own bias speaking :P

**Comment by edwardloveall** [comment]
I love component classes. That makes a ton of sense to me.

I'm not a big fan of `responds_to?` I think because it's a conditional where as the component classes could use polymorphism and keep pages much cleaner. I've seen conditionals get way out of hand views for other apps so I find myself trying to avoid them.

Example

```crystal
render do
  UserMessage.for(@user).messages.each do |message|
    # ...
  end
end

class UserMessage
  def self.for(user)
    return BannedMessage.new(user) if user.banned?
    return WarningMessage.new(user) if user.has_active_warnings?
    return PremiumMessage.new(user) if uesr.active_account?
    NullMessage.new
  end
end

class BannedMessage
  def messages
    user.flags.warning.messages
  end
end # etc...
```

vs 

```crystal
render do
  if user.responds_to?(:banned_message)
    render_banned_message
  elsif user.responds_to?(:warning_messages)
    render_warning_messages
  elsif # etc...
  end
end
```

I keep changing examples on you, which isn't fair 😄. In general, my philosophy is don't ask if the class can do something, just tell it and it will respond appropriately.

**Comment by edwardloveall** [comment]
This might be getting too wrapped up in a higher level "What do views look like for lucky" rather than addressing this particular issue. I say go for your original idea (it's a good one), and we can open other issues/talk on gitter to figure out alternate solutions.

**Comment by satb** [comment]
@paulcsmith - When PageWithSidebar defines the abstract methods, does that mean all pages have to define sidebar and inner? And if most, if not all, pages have the same sidebar and inner, they are still forced to define those in every page or could those be extracted out into a separate component? I'd be careful not to push inheritance hierarchy too much and favor composition instead. In that context, we could consider the overall outline of the web page as the "layout" which is "composed" of other pieces. Like edwardlowall says, components and composition without too much responds_to sprinkled all over will be great.

**Comment by satb** [comment]
Thanks. When you say "_You'd still have a MainPage for the other pages. Just like you might have an AuthPage that you use for password resets, sign in, sign up, etc._", I take that to mean the MainLayout/LuckyWeb::Layout will also be around because MainPage subclasses use the MainLayout. But the issue was opened by saying layouts are not required. Thats where I am getting lost I guess. The issue was opened by saying Layouts are not required (might be removed from future versions of Lucky?), but updates to the issue say MainPage will still be there but MainPage currently relies on MainLayout.

**Comment by paulcsmith** [comment]
@satb I can see why what I'm saying is confusing. You would still have layouts in terms of *design*. Like you might have one page layout that looks very different than another page looks like. One might have nothing but a sign up form, and another might head a logo, header, navbar and a footer. The way Lucky works right now, you'd create a new Layout class like `AuthLayout` and a new page `AuthPage` that you then inherit from.

This proposal is to remove the idea of separate Layout classes (we'd still have different design layouts), but they'd be implemented using a base page instead of a separate class. So instead of `AuthLayout` and `AuthPage` you'd have just `AuthPage` that you inherit from. It's simplifying the code from having 2 objects to having just one. Does that make more sense now?

**Comment by satb** [comment]
Got it. Thanks. It makes much more sense now why you proposed removing Layout. Its more per-page Layout that is being removed.

**Comment by paulcsmith** [comment]
Part of the problem is likely that I'm oversimplifying examples and rushing an explanation. But here's an example of how I would do two base pages for two differently *designed* page layouts

https://gist.github.com/paulcsmith/60cf938e65c4a5728208a60f3b52af20

```crystal
abstract class AuthPage < BasePage
  render do
    html do
      head do
        # title, css, etc.
      end
      
      body do
        div class: "auth-form-title" do
          form_title
        end
        
        div class: "auth-form-body" do
          inner
        end
      end
    end
  end
  
  abstract def form_title
  abstract def inner
end

abstract class PageWithSidebar < BasePage
  render do
    html do
      head do
        # title, css, etc.
      end
      
      body do
        div class: "breadcrumbs" do
          breadcrumbs
        end
        
        div class: "main-content" do
          # Child classes can optionally show help text on the page
          if responds_to?(:help_text)
            div class: "help-text" do
              self.help_text
            end
          end
        
          inner
        end
        
        div class: "sidebar" do
          sidebar
        end
      end
    end
  end

  abstract def sidebar
  abstract def inner
  abstract def breadcrumbs
  # Note there is no abstract def for help_text because only some child pages may want it
end
```

**Comment by edwardloveall** [comment]
Yeah, I think this is good. It looks like we're all in agreement that we don't need a layout class, only abstract page classes. That feels good to me.

I think I got us off track talking about `responds_to?` vs components. Sorry about that. I think it may end up being a stylistic choice that we can flesh it out more in a separate discussion if we want to. 😄 👍

**Comment by paulcsmith** [comment]
Ok cool. I'll work on this sometime this week or next. I'm also thinking of naming the abstract classes "layouts" so that's it is easier to talk about them. "Use the SidebarLayout for that page" or "Let's make a new AuthLayout that we can use for our sign up pages." Seems minor, but I think the distinction might help when people talk and write about these concepts.

```crystal
class SignIns::NewPage < AuthLayout # Rather than inheriting from AuthPage
end
```

Thoughts?

**Comment by satb** [comment]
I really like the idea of renaming it to Layout. It conveys the semantics and the intent more clearly. I was to ask the same question as to why it is being called Page then, but you have the answer above already 👍

**Comment by edwardloveall** [comment]
Yeah, I like that too. Along with what @satb and @paulcsmith said about semantics, it will immediately make sense coming from a rails context.

So this kind of becomes "Remove LuckyWeb::Page in favor of just Layout"?

**Comment by paulcsmith** [comment]
@edwardloveall Good point. I think that whatever the class/module is named, it should work for using in layouts and also plain pages if you don't need a layout. Just so it makes sense. I'm thinking `LuckyWeb::Html`. I think this would make sense if used in a layout or directly on a page that doesn't inherited from a layout.

```crystal 
abstract class MainLayout
  # Imports tags, rendering, needs, etc.
  include LuckyWeb::Html
end

class SimpleErrorPageWithoutLayout
  include LuckyWeb::Html
end
```

**Comment by edwardloveall** [comment]
`Lucky::Html` makes perfect sense. And layout is a good name for the abstract class.

**Comment by paulcsmith** [comment]
Right on. Thanks for talking this through with me @satb @SViccari and @edwardloveall. It's been incredibly helpful

**Comment by 5minpause** [comment]
I am a bit late to this show, but it really does make sense to remove the page and just keep the layout.
